crates/indexer/src/indexer/fetcher.rs

```diff
@@ -0,0 +1,455 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::counters::{FETCHED_TRANSACTION, UNABLE_TO_FETCH_TRANSACTION};
+use aptos_api::Context;
+use aptos_api_types::{AsConverter, LedgerInfo, Transaction, TransactionOnChainData};
+use aptos_logger::prelude::*;
+use aptos_vm::data_cache::StorageAdapterOwned;
+use futures::channel::mpsc;
+use futures::{SinkExt, StreamExt};
+use serde_json::Value;
+use std::sync::Arc;
+use std::time::Duration;
+use storage_interface::state_view::DbStateView;
+use tokio::task::JoinHandle;
+
+// Default Values
+const RETRY_TIME_MILLIS: u64 = 300;
+const MAX_RETRY_TIME_MILLIS: u64 = 120000;
+const TRANSACTION_FETCH_BATCH_SIZE: u16 = 500;
+const TRANSACTION_CHANNEL_SIZE: usize = 35;
+
+#[derive(Debug)]
+pub struct Fetcher {
+    pub context: Arc<Context>,
+    options: TransactionFetcherOptions,
+    chain_id: u8,
+    current_version: u64,
+    highest_known_version: u64,
+    transactions_sender: mpsc::Sender<Vec<Transaction>>,
+}
+
+impl Fetcher {
+    pub fn new(
+        context: Arc<Context>,
+        starting_version: u64,
+        options: TransactionFetcherOptions,
+        transactions_sender: mpsc::Sender<Vec<Transaction>>,
+    ) -> Self {
+        Self {
+            context,
+            options,
+            chain_id: 0,
+            current_version: starting_version,
+            highest_known_version: starting_version,
+            transactions_sender,
+        }
+    }
+
+    pub fn set_highest_known_version(&mut self) -> anyhow::Result<()> {
+        let info = self.context.get_latest_ledger_info_wrapped()?;
+        self.highest_known_version = info.ledger_version.0;
+        self.chain_id = info.chain_id;
+        Ok(())
+    }
+
+    pub async fn run(&mut self) {
```

Comment:
could you write the algorithm here?

it looks like you grab the latest version
then you find out how many chunks to get, then you spawn tasks for each chunk, and wait?